### PR TITLE
Add an e2e test for writing an Iceberg table with pyiceberg and reading it with DataFusion

### PR DESCRIPTION
**Pull request by gruuya**
Besides serving as an e2e test, this also excessive quite a bit of type conversion logic.

In addition it demonstrates the issue described in https://github.com/apache/iceberg-rust/issues/813.

### CODE REVIEW

File: crates/integration_tests/testdata/pyiceberg/load_types_table.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from datafusion import SessionContext
+from pyiceberg.catalog import load_catalog
+import pyarrow.parquet as pq
+
+# Generate a table with various types in memory and dump to a Parquet file
+ctx = SessionContext()
```
Comment (Fokko) [inline crates/integration_tests/testdata/pyiceberg/load_types_table.py:23]: I'm reluctant to let Datafusion generate these files, for two reasons:

- These files are imported into an Iceberg table and are missing things like Field-IDs etc.
- I'd rather depend on purely Spark which uses the Java SDK underneath, which is more complete than PyIceberg. The Java impl is considered the reference implementation.

Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py:24]: > missing things like Field-IDs etc.

If by that you mean `PARQUET:field_id` note that pyiceberg actualy decorates the fields during table creation with those, I've just pushed an [addendum](https://github.com/apache/iceberg-rust/pull/825/commits/b5d23ffcc8f3998ee16df1dde30a0bfc84173b0a) to the tests that demonstrate this.

Granted I'm not sure about the state of pyiceberg compliance with the protocol.

Would you be more open to accepting this test if I switched to Java SDK for generating the Parquet files and creating the table?

Comment (Fokko) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py:24]: I also see value in this, let's hear what others think!

File: crates/integration_tests/testdata/pyiceberg/load_types_table.py
Comment (kevinjqliu) [inline crates/integration_tests/testdata/pyiceberg/load_types_table.py:1]: nit: wdyt of renaming this `provision.py` mirroring what we do for the spark setup 
https://github.com/apache/iceberg-rust/blob/main/crates/integration_tests/testdata/spark/provision.py

File: crates/integration_tests/testdata/pyiceberg/load_types_table.py (hunk 1 above)
Comment (kevinjqliu) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py:24]: Generally i like the idea of using `pyiceberg` for testing, maybe also  alongside spark. This can help facilitate cross platform interoperability, 
but im not sure when we would use one versus the other.

File: crates/integration_tests/testdata/pyiceberg/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM python:3.9-bullseye
+
+RUN pip install pyiceberg==0.8 pyarrow==18.0 datafusion==43.1.0
```
Comment (kevinjqliu) [inline crates/integration_tests/testdata/pyiceberg/Dockerfile:18]: nit: for `pyarrow`, use `pyiceberg[pyarrow]` instead.

File: crates/integration_tests/testdata/pyiceberg/load_types_table.py (hunk 1 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py:24]: Alright just an update—I've tried doing the same with pyspark but it seems it is being very conservative and either rejecting some types outright (e.g. `Illegal Parquet type: INT64 (TIMESTAMP(NANOS,false))`), or coercing them on its own.

For instance if I just supply it with a Parquet file like `type_test.parquet` (output of `load_types_table.py`) with type hints
```
message arrow_schema {
  optional boolean cboolean;
  optional int32 cint8 (INTEGER(8,true));
  optional int32 cint16 (INTEGER(16,true));
  optional int32 cint32;
  optional int64 cint64;
...
```
and do
```python
parquet_df = spark.read.parquet("types_test.parquet")
parquet_df.writeTo("rest.default.types_test").using("iceberg").create()
```
the resulting data parquet files have incompatible type hints removed
```
message table {
  optional boolean cboolean = 1;
  optional int32 cint8 = 2;
  optional int32 cint16 = 3;
  optional int32 cint32 = 4;
  optional int64 cint64 = 5;
...
```

TLDR: I can't seem to replicate the issue in #814 with pyspark, though it could still be used to test out the non-corner cases.

File: crates/integration_tests/testdata/pyiceberg/Dockerfile (hunk 2 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/Dockerfile]: good point, revised!

File: crates/integration_tests/testdata/pyiceberg/load_types_table.py (hunk 1 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py]: > but im not sure when we would use one versus the other.

I think there's definitely merit in having an integration types test like the one proposed here. The basic one can be done with pyspark, though it might be a bonus (or not) that apparently pyiceberg can hit some corner cases which pyspark can't.

(fwiw, this is something that we ran into in our dev work, so it's not a end-user-reported issue.)

Comment (kevinjqliu) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py]: > apparently pyiceberg can hit some corner cases which pyspark can't.

Interesting! curious to hear about what you guys ran into. 

I think Pyspark is good for integration tests through the spark API. PyIceberg allows more granular test cases since you can work with iceberg constructs

Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py]: > curious to hear about what you guys ran into.

Actually the only one so far seems to be this Iceberg-from-parquet creation discussed above—if the Parquet file has e.g. Int32 field with Int16 type hints spark will automatically coerce those in the resulting data file into Int32, whereas pyiceberg will leave it as is (and so during the scan in the actual record batches we get Int16, whilst the schema says it's Int32).

Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/load_types_table.py]: Ok, I've opened an alternative PR that adds an integration test starting off from spark in #850.

Granted this doesn't hit the original issue #813, but could still be worthwhile.

File: crates/integration_tests/testdata/pyiceberg/provision.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+import os
+from datafusion import SessionContext
+from pyiceberg.catalog import load_catalog
+import pyarrow.parquet as pq
+
+# Generate a table with various types in memory and dump to a Parquet file
+ctx = SessionContext()
+ctx.sql("""
+CREATE TABLE types_test (
```
Comment (kevinjqliu) [inline crates/integration_tests/testdata/pyiceberg/provision.py:26]: lets generate these values in pyiceberg/pyarrow. it'll help us figure out the supportability of each types which is where the E2E test is useful.

Here are the supported types we use in pyiceberg's integration test 
https://github.com/apache/iceberg-python/blob/acd6f5a8a19db709e835e2686b87d4db3dca254f/tests/conftest.py#L304-L349

**Review by kevinjqliu — COMMENTED** [review]
I think E2E test should be pyiceberg writes and reading with datafusion. similar to #850

other than that, generally lgtm!

File: crates/integration_tests/testdata/pyiceberg/provision.py (hunk 3 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/provision.py]: Sounds good, I just switched to pyarrow to generate the Parquet file from which the table is constructed.

To be clear, if #850 gets accepted and merged, and if folks don't feel comfortable including pyiceberg in the testing scope for now, then this PR can just remain as an example for the issue in #813 (it still exercises it as demonstrated at the very end of the integration test).

File: crates/integration_tests/testdata/pyiceberg/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM python:3.9-bullseye
+
+RUN pip install pyiceberg[pyarrow]==0.8
```
Comment (kevinjqliu) [inline crates/integration_tests/testdata/pyiceberg/Dockerfile:18]: nit: 0.8.1 is the latest version

File: crates/integration_tests/testdata/pyiceberg/provision.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        "s3.secret-access-key": os.environ["AWS_SECRET_ACCESS_KEY"],
+    },
+)
+
+# Create a corresponding Iceberg table and append the file to it
+iceberg_table = catalog.create_table_if_not_exists(
+    identifier=f"default.types_test",
+    schema=parquet_file.schema,
+)
+iceberg_table.append(df=parquet_file)
```
Comment (kevinjqliu) [inline crates/integration_tests/testdata/pyiceberg/provision.py:87]: `iceberg_table.append(table)` also works, do we need to roundtrip writing to parquet and then reading it back?

File: crates/integration_tests/testdata/pyiceberg/provision.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    ('cint64', pa.int64()),
+    ('cfloat32', pa.float32()),
+    ('cfloat64', pa.float64()),
+    ('cdecimal128', pa.decimal128(8, 2)),
+    ('cdate32', pa.date32()),
+    ('ctimestamp', pa.timestamp('us')),
+    ('ctimestamptz', pa.timestamp('us', tz='UTC')),
+    ('cutf8', pa.utf8()),
+    ('cbinary', pa.binary()),
+])
```
Comment (kevinjqliu) [inline crates/integration_tests/testdata/pyiceberg/provision.py:55]: do we want to include other data types? including complex types? like in https://github.com/apache/iceberg-python/blob/acd6f5a8a19db709e835e2686b87d4db3dca254f/tests/conftest.py#L304-L349

**Review by kevinjqliu — APPROVED** [review]
LGTM! I like this E2E test from pyiceberg writes to datafusion read. 
In the future, I want to expose the IcebergTableProvider as python binding so we can also use the datafusion python library to query (#865)

File: crates/integration_tests/testdata/pyiceberg/provision.py (hunk 5 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/provision.py]: Oh I thought only appending via a parquet file reproduces #813, but turns out so does appending a pyarrow table directly; nice catch!

File: crates/integration_tests/testdata/pyiceberg/provision.py (hunk 6 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/provision.py:54]: Makes sense to me, could be valuable to include list/map/struct.

**Review by kevinjqliu — DISMISSED** [review]
LGTM!

File: crates/integration_tests/tests/datafusion.rs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        "| cboolean | cint8 | cint16 | cint32 | cint64 | cfloat32 | cfloat64 | cdecimal128 | cdate32    | ctimestamp          | ctimestamptz         | cutf8 | cbinary |",
+        "+----------+-------+--------+--------+--------+----------+----------+-------------+------------+---------------------+----------------------+-------+---------+",
+        "| false    | -128  | 0      | 0      | 0      | 0.0      | 0.0      | 0.00        | 1970-01-01 | 1970-01-01T00:00:00 | 1970-01-01T00:00:00Z | 0     | 30      |",
+        "| true     | -127  | 1      | 1      | 1      | 1.0      | 1.0      | 0.01        | 1970-01-02 | 1970-01-01T00:00:01 | 1970-01-01T00:00:01Z | 1     | 31      |",
+        "| false    | -126  | 2      | 2      | 2      | 2.0      | 2.0      | 0.02        | 1970-01-03 | 1970-01-01T00:00:02 | 1970-01-01T00:00:02Z | 2     | 32      |",
+        "+----------+-------+--------+--------+--------+----------+----------+-------------+------------+---------------------+----------------------+-------+---------+",
+    ];
+    assert_batches_eq!(expected, &batches);
+
+    // TODO: this isn't OK, and should be fixed with https://github.com/apache/iceberg-rust/issues/813
```
Comment (kevinjqliu) [inline crates/integration_tests/tests/datafusion.rs:138]: should we wait for this fix?

Comment (Xuanwo) [reply crates/integration_tests/tests/datafusion.rs:138]: I think we can move first.

**Review by Xuanwo — APPROVED** [review]
Thank you @gruuya for working on this.

**Comment by Xuanwo** [comment]
Hi, sorry for the late merge call. Would you like to help fix the build? @gruuya

**Comment by gruuya** [comment]
> Hi, sorry for the late merge call. Would you like to help fix the build? @gruuya

Hey, no worries! I think the build is good now, thanks.

On the other hand, one form of the types integration test was already accepted in #850 (spark->DataFusion). I also know that @Fokko had some reservations about including pyiceberg in test coverage for now.

So the only merit for merging this would be exercising #813 it seems (which I think is fine if everyone is aligned on pyiceberg).

File: crates/integration_tests/testdata/pyiceberg/provision.py (hunk 6 above)
Comment (gruuya) [reply crates/integration_tests/testdata/pyiceberg/provision.py:56]: Added a list as well; with map/struct the tests seems to become too verbose, but I can add those too if people don't mind.
